Added OperatorMap for mapping Operator to any template <T>

edvgha opened the pull request:
Summary: Generic way to check if Operator belongs to predefined map, and if so via public method(s) access to map value. In general value can be anything for example Operator's schema.

Test Plan: buck test caffe2/test/cpp/jit:jit -- OperatorMap

Differential Revision: D28357933

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1518,68 @@ struct OperatorSet {
+  }
+
+ private:
+  std::unordered_map<Symbol, T> map;
```
In `OperatorSet` and `OperatorMap` we will have multiple schemas with the same node kind. For example:
```
      "aten::mul.Tensor(Tensor self, Tensor other) -> Tensor",
      "aten::mul.Scalar(Tensor self, Scalar other) -> Tensor",
```

I think, as in OperatorSet, we want to map from Symbol to a vector, and then within the vector store pairs of `std::shared_ptr<Operator>, T`

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1518,68 @@ struct OperatorSet {
+    }
+  }
+
+  void erase(const Symbol& key) {
```
I think the public APIs should only take in `Operator` as keys, and not Symbols/strings (related to the other comment).

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1518,68 @@ struct OperatorSet {
+    map.erase(Symbol::fromQualString(key));
+  }
+
+  T find_or(const Symbol& key, T t = T()) {
```
nit: i would maybe leave this off the first API.

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1518,68 @@ struct OperatorSet {
 };
 
+template <typename T>
+struct OperatorMap {
```
Nit: maybe move more of the implementation to `ir.cpp` following along with convention for `OperatorSet`

eellison reviewed (commented):
Looks good! I think we still need to handle the case where multiple operators are registered and their symbols collide.

edvgha commented:
I'll suggest some small redesign, move Note::matches into Operator.
First of all it is what exactly the method is doing (matching Operator's schemas) and in this case we can add more functionality into OperatorSet and OperatorMap without code duplication.

edvgha commented:
I was hurry, in Node::matches we are try to match Node's kind()/inputs() with Operator's schema() rather than Node's op_ with Operator's schema(), since during the call Node's op_ might not be valid at all.

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1527,63 @@ struct OperatorSet {
+    return map.count(Symbol::fromQualString(op.schema().name())) != 0;
+  }
+
+  c10::optional<ValueType> find(const Operator& op) {
```
Returning `ValueType` breaks the abstraction that we're storing Operators and mapping them to `T`. We should be returning `c10::optional<T>` here

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1527,63 @@ struct OperatorSet {
+    map.erase(Symbol::fromQualString(op->schema().name()));
+  }
+
+  bool contains(const Operator& op) const {
```
this is checking the QualString but we should also be checking that one of the Operators stored within the vector matches

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1527,63 @@ struct OperatorSet {
+    return map.size();
+  }
+
+  int64_t value_size(const std::shared_ptr<Operator>& op) const {
```
i would leave off value_size as part of the public api because it's kind of confusing what it does and breaks the `Operator->T` abstraction

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1527,63 @@ struct OperatorSet {
+  }
+
+  void erase_all(const std::shared_ptr<Operator>& op) {
+    map.erase(Symbol::fromQualString(op->schema().name()));
```
this should be removing from the vector and then the map if the vector is size 0, right now we are removing unrelated operators with same qualstring

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1527,63 @@ struct OperatorSet {
+    }
+  }
+
+  void erase_all(const std::shared_ptr<Operator>& op) {
```
if we're going to offer `erase` as part of public API, i think it should just be `erase`

eellison reviewed (commented):
Looks good, a few more comment

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1512,5 +1527,77 @@ struct OperatorSet {
+  }
+
+  void insert(const std::shared_ptr<Operator>& op, T val) {
+    map[Symbol::fromQualString(op->schema().name())].emplace_back(
```
should we enforce no duplicates ? the existing `OperatorSet` doesn't do it but that's because it's a set and not a map... Maybe we should erase and then insert here ?

eellison reviewed (approved):
Looks great!! Thanks for doing this!!

edvgha committed:
Added OperatorMap for mapping Operator to any template <T> (#58060)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/58060

Generic way to check if Operator belongs to predefined map, and if so via public method(s) access to map value. In general value can be anything for example Operator's schema.

Test Plan: buck test caffe2/test/cpp/jit:jit -- OperatorMap

Reviewed By: Krovatkin

Differential Revision: D28357933

fbshipit-source-id: bff0ef64ced00038263fe7e3e06702dd982eb444
